Add cuda version of SpatialBNOp also optimize SpatialBN on CPU

xiaomengy opened the pull request:
Summary: Add cuda version of SpatialBNOp also optimize SpatialBN on CPU

Differential Revision: D9512435

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

houseroad commented on caffe2/operators/spatial_batch_norm_op.h:
```diff
@@ -11,69 +16,377 @@ template <class Context>
+        math::Set<T, Context>(C, T(0), saved_rstd_data, &context_);
+        return true;
+      }
+      if (num_batches_ > 1) {
```
Add some comments about the computation equation here?

houseroad commented on caffe2/operators/spatial_batch_norm_op.h:
```diff
@@ -11,69 +16,377 @@ template <class Context>
  protected:
-  bool is_test_;
+  template <typename T>
+  void ComputeFusedParam(
```
why we call it fusedParam?

BIT-silence replied:
for the Forward pass, the actual formula is Y = (X - mu) / std * scale + bias. In order to reduce the number of computation, we fused it to alpha = scale / std, beta = bias - mu / std * scale. Then the formula is Y = alpha * X + beta. Here we are computing alpha and beta which is the fused param for the forward pass.

houseroad commented on caffe2/operators/spatial_batch_norm_op.h:
```diff
@@ -11,69 +16,377 @@ template <class Context>
+
+  template <typename T>
+  bool DoRunWithType() {
+    const auto& X = Input(INPUT);
```
Let's just keep the old comments about how to compute the gradients.

houseroad reviewed (approved):
LG, please add some comments about the computation.

bddppq commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

BIT-silence commented:
@pytorchbot retest this please

xiaomengy committed:
Add cuda version of SpatialBNOp also optimize SpatialBN on CPU (#10888)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/10888

Add cuda version of SpatialBNOp also optimize SpatialBN on CPU

Reviewed By: houseroad

Differential Revision: D9512435

fbshipit-source-id: 6e5f2eeb420ec2831f473972a614677e8dfaf906

BIT-silence commented:
@pytorchbot retest this please
